core/services/ocr2/plugins/threshold/decryption_queue.go

```diff
@@ -0,0 +1,236 @@
+package threshold
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/smartcontractkit/chainlink/v2/core/logger"
+	"github.com/smartcontractkit/chainlink/v2/core/services/job"
+)
+
+type CiphertextId = []byte
+
+type Decryptor interface {
+	Decrypt(ctx context.Context, ciphertextId CiphertextId, ciphertext []byte) ([]byte, error)
+}
+
+// This interface will be replaced by the Threshold Decryption Plugin when it is ready
+type DecryptionQueuingService interface {
+	GetRequests(requestCountLimit int, totalBytesLimit int) []DecryptionRequest
+	GetCiphertext(ciphertextId CiphertextId) ([]byte, error)
+	ReturnResult(ciphertextId CiphertextId, plaintext []byte)
+}
+
+type DecryptionRequest struct {
+	ciphertextId CiphertextId
+	ciphertext   []byte
+}
+
+type pendingRequest struct {
+	chPlaintext chan<- []byte
+	ciphertext  []byte
+}
+
+type completedRequest struct {
+	plaintext []byte
+	timer     *time.Timer
+}
+
+type decryptionQueue struct {
+	maxQueueLength                int
+	maxCiphertextBytes            int
+	completedRequestsCacheTimeout time.Duration
+	pendingRequestQueue           []CiphertextId
+	pendingRequests               map[string]pendingRequest
+	completedRequests             map[string]completedRequest
+	mu                            sync.RWMutex
+	lggr                          logger.Logger
+}
+
+var (
+	_ Decryptor                = &decryptionQueue{}
+	_ DecryptionQueuingService = &decryptionQueue{}
+	_ job.ServiceCtx           = &decryptionQueue{}
+)
+
+func NewDecryptionQueue(maxQueueLength int, maxCiphertextBytes int, completedRequestsCacheTimeout time.Duration, lggr logger.Logger) *decryptionQueue {
+	dq := decryptionQueue{
+		maxQueueLength,
+		maxCiphertextBytes,
+		completedRequestsCacheTimeout,
+		[]CiphertextId{},
+		make(map[string]pendingRequest),
+		make(map[string]completedRequest),
+		sync.RWMutex{},
+		lggr.Named("decryptionQueue"),
+	}
+	return &dq
+}
+
+func (dq *decryptionQueue) Decrypt(ctx context.Context, ciphertextId CiphertextId, ciphertext []byte) ([]byte, error) {
+	if len(ciphertext) > dq.maxCiphertextBytes {
+		return nil, errors.New("ciphertext too large")
+	}
+
+	chPlaintext, err := dq.getResult(ciphertextId, ciphertext)
+	if err != nil {
+		return nil, err
+	}
+
+	select {
+	case pt, ok := <-chPlaintext:
+		if ok {
+			return pt, nil
+		}
+		return nil, fmt.Errorf("pending decryption request for ciphertextId %s was closed without a response", ciphertextId)
+	case <-ctx.Done():
+		dq.mu.Lock()
+		defer dq.mu.Unlock()
+		delete(dq.pendingRequests, string(ciphertextId))
```

Comment:
I see you always cast this to string.
Wouldn't it be more convenient to have:
`type CiphertextId = string`

Reply:
I almost suggested the same, but FWIW these map key uses are optimized by the compiler to avoid the usual copy, so this is idiomatic provided that `[]byte` makes sense the rest of the time :shrug: 
https://github.com/golang/go/issues/3512